### ISSUE TITLE
Wishlist: Multidimensional colormaps

### ISSUE THREAD

**Issue by ischwabacher** [matplotlib/matplotlib#4369]
xref #3343

TL;DR: It would be nice if it were possible to have color maps for multidimensional data, instead of expecting the color to depend only on a single quantity as is currently the case.

I'm currently working on an application that would like to be able to modulate the color and opacity of an image independently (as in Figure 3.B.b of [Allen et al. 2012](http://www.sciencedirect.com/science/article/pii/S089662731200428X)) so that color indicates a parameter estimate and opacity indicates a statistic assessing the reliability of that estimate:

![Allen et al. 2012, Fig. 3](http://ars.els-cdn.com/content/image/1-s2.0-S089662731200428X-gr3.jpg)

Currently, this must be accomplished by applying a color map to the parameter estimate, then overwriting the alpha channel with the statistical map, and finally calling `imshow` with the computed RGBA data.  (This is what @mwaskom is doing in the referenced issue.)  This is a nuisance, and more importantly, it forces all interpolation to be done in RGBA space instead of in data space.  (I can't find the xref for it, but I believe there is an open issue to make `imshow` interpolation act on the input array instead of the color-mapped array.)

Rather than simply making it possible to control alpha independently of RGB, it would make sense to allow color maps that map multidimensional data to RGBA space.  Then `imshow`'s color map for 3-dimensional input can default to *id &oplus; 1*, and for 4-dimensional input to *id*, which gets rid of those special cases and makes it possible to have other color maps of those dimensionalities.

One could imagine other use cases for this in neuroimaging; for instance, [diffusion tensor images](http://en.wikipedia.org/wiki/Diffusion_MRI) are commonly displayed with hue depending on the principal eigenvector of a symmetric tensor field, while luminosity depends on the dispersion of the eigenvalues.

![Diffusion Tensor image](http://fsl.fmrib.ox.ac.uk/fsl/fslview/fslview_dti.png)

In a completely different direction, one could display [a complex-valued function](http://en.wikipedia.org/wiki/Theta_function) by splitting the complex data into real and imaginary parts, and then applying a two-dimensional color map that sets hue based on the argument and luminosity based on the modulus of the input:

![The Jacobi theta function](http://upload.wikimedia.org/wikipedia/commons/8/85/Complex_theta_minus0point1times_e_i_pi_0point1.jpg)

**Comment by tacaswell** [comment]
All sounds great to me.  I see no technical issue, it is just a matter of someone doing this.

If we only want to map 2D data -> rgba than I think a family of normalizers which go from data -> unit-disk (or is unit square more sensible) in the complex plane and then color maps that go from the unit disk -> RGBA would be a good start.  

If we want to do higher dimensions map to quaternions (or maybe that is getting too fancy?).

**Comment by OceanWolf** [comment]
I guess though we shouldn't limit ourselves to 2d colour spaces.

Thinking of colour spaces, have you heard of the L*a*b* colour space?  http://en.wikipedia.org/wiki/Lab_color_space as it "aspires to perceptual uniformity" I wonder if this could help influence our choice of colourmaps for our colour revolution...

**Comment by OceanWolf** [comment]
Some other related issues:
* The documentation on imshow says "MxN – luminance (grayscale, float array only)" even though it doesn't relate to luminance, it maps the value to the (specified/default) colourmap.
* When we plot a MxNx3 and MxNx4 array, we can still add a colorbar despite it having no meaning to the data.

```python
import numpy as np
import matplotlib.pyplot as plt

data = np.random.rand(6,5,3)
plt.imshow(data, interpolation='none')
plt.colorbar()
plt.show()
```

If we wanted to plot multi-dimensional colormaps then I guess we need to change the colorbar implementation as well to deal with this...

**Comment by tacaswell** [comment]
The 'luminescence' in this case is a synonym for monochrome (which coming from a light microscopy background means luminescence on the detector side) so the documentation is fine.

**Comment by endolith** [comment]
See also [discussion on Stack Overflow](http://stackoverflow.com/q/15207255/125507).

[I've been doing this "by hand" in Lab space](https://www.flickr.com/photos/56868697@N00/albums/72157656610100923) with `mpmath.cplot`, which takes a function to generate the RGB color from the amplitude and phase of each sample (but is extremely slow):

![5th order Bessel filter](https://c2.staticflickr.com/2/1511/24508017860_232e5ec541_o.png)

**Comment by sbenthall** [comment]
+1 wanting this feature.

**Comment by efiring** [comment]
@njsmith, since this involves color spaces and targets neuroscience, perhaps among other applications, do you have some advice?

**Comment by efiring** [comment]
The simplest version of this, via control of alpha, is being revived in #6268.
My suspicion is that perhaps anything beyond this should be handled as a separate package that pre-processes the inputs and returns RGBA image or QuadMesh arrays for display by matplotlib's existing functionality, or that invokes that functionality directly.

**Comment by arshadpatel2001** [comment]
Hey, is anyone working on this?
If not I would like to contribute towards the same And would like to know more about it, and how I can get started?

**Comment by OceanWolf** [comment]
@arshadpatel2001 Have you taken a look at the repository of @endolith posted above, that sounds like a good first step.  I don't know if we need to extend the concept of perceptually-uniform to 2d, i.e. if so maybe we take the | grad C | to create a perceptual scalar field?

**Comment by timhoffm** [comment]
Tracked in #14168, and worked on in #28658, #28454, #29221.
